FixedQParamsFakeQuantize: adjust default quant_min and quant_max

vkuzo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#47423 FixedQParamsFakeQuantize: adjust default quant_min and quant_max**

Summary:

Since the dtype of this fake_quant is `quint8`, the output range should be
from 0 to 255.  Fixing.  This should address the numerical inaccuracies with
sigmoid and hardsigmoid with `FixedQParamsFakeQuantize` attached compared
to their quantized counterparts.

In a future PR, might be safer to also make the activation functions
using `FixedQParamsFakeQuantize` to explicitly specify their expected
output range and zero_point.  Leaving that for later, as this bugfix
should be landed urgently.

Test Plan:

Manual script which gives low SQNR before this PR and high SQNR after
this PR: https://gist.github.com/vkuzo/9906bae29223da72b10d6b6aafadba42

```
python test/test_quantization.py TestEagerModeQATOps
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D24751497](https://our.internmc.facebook.com/intern/diff/D24751497)

vkuzo committed:
FixedQParamsFakeQuantize: adjust default quant_min and quant_max

Summary:

Since the dtype of this fake_quant is `quint8`, the output range should be
from 0 to 255.  Fixing.  This should address the numerical inaccuracies with
sigmoid and hardsigmoid with `FixedQParamsFakeQuantize` attached compared
to their quantized counterparts.

In a future PR, might be safer to also make the activation functions
using `FixedQParamsFakeQuantize` to explicitly specify their expected
output range and zero_point.  Leaving that for later, as this bugfix
should be landed urgently.

Test Plan:

Manual script which gives low SQNR before this PR and high SQNR after
this PR: https://gist.github.com/vkuzo/9906bae29223da72b10d6b6aafadba42

https://github.com/pytorch/pytorch/pull/47376, which can be landed after
this, adds a proper test.

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 reviewed (approved):
oh..thanks for the fix!

jerryzh168 commented on torch/quantization/fake_quantize.py:
```diff
@@ -243,10 +243,13 @@ def extra_repr(self):
 default_weight_fake_quant = FakeQuantize.with_args(observer=MovingAverageMinMaxObserver, quant_min=-128, quant_max=127,
                                                    dtype=torch.qint8, qscheme=torch.per_tensor_symmetric, reduce_range=False)
+
+# TODO(future PR): remove these defaults and enforce activation functions
```
I think it's fine to have defaults since these are added automatically for certain operators, theses are not configured by user.

vkuzo replied:
yeah, it's not technically related to this PR, more of a code style.  We can discuss at a later time, I don't feel too strongly.

jerryzh168 commented on torch/quantization/fake_quantize.py:
```diff
@@ -243,10 +243,13 @@ def extra_repr(self):
+
+# TODO(future PR): remove these defaults and enforce activation functions
+# to explicitly specify their output range
 default_symmetric_fixed_qparams_fake_quant = FixedQParamsFakeQuantize.with_args(
```
should we explicitly specify quant_min/quant_max for this one as well?

vkuzo committed:
add test on "FixedQParamsFakeQuantize: adjust default quant_min and quant_max"


Summary:

Since the dtype of this fake_quant is `quint8`, the output range should be
from 0 to 255.  Fixing.  This should address the numerical inaccuracies with
sigmoid and hardsigmoid with `FixedQParamsFakeQuantize` attached compared
to their quantized counterparts.

In a future PR, might be safer to also make the activation functions
using `FixedQParamsFakeQuantize` to explicitly specify their expected
output range and zero_point.  Leaving that for later, as this bugfix
should be landed urgently.

Test Plan:

Manual script which gives low SQNR before this PR and high SQNR after
this PR: https://gist.github.com/vkuzo/9906bae29223da72b10d6b6aafadba42

https://github.com/pytorch/pytorch/pull/47376, which can be landed after
this, adds a proper test.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D24751497](https://our.internmc.facebook.com/intern/diff/D24751497)

[ghstack-poisoned]

vkuzo commented:
@jerryzh168 , I pulled some of https://github.com/pytorch/pytorch/pull/47376 in here, just so the test is in the same PR, hope that's ok!  I didn't add the leaky_relu part from that PR.

vkuzo committed:
comments on "FixedQParamsFakeQuantize: adjust default quant_min and quant_max"


Summary:

Since the dtype of this fake_quant is `quint8`, the output range should be
from 0 to 255.  Fixing.  This should address the numerical inaccuracies with
sigmoid and hardsigmoid with `FixedQParamsFakeQuantize` attached compared
to their quantized counterparts.

In a future PR, might be safer to also make the activation functions
using `FixedQParamsFakeQuantize` to explicitly specify their expected
output range and zero_point.  Leaving that for later, as this bugfix
should be landed urgently.

Test Plan:

Manual script which gives low SQNR before this PR and high SQNR after
this PR: https://gist.github.com/vkuzo/9906bae29223da72b10d6b6aafadba42

https://github.com/pytorch/pytorch/pull/47376, which can be landed after
this, adds a proper test.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D24751497](https://our.internmc.facebook.com/intern/diff/D24751497)

[ghstack-poisoned]

vkuzo committed:
rebase on "FixedQParamsFakeQuantize: adjust default quant_min and quant_max"


Summary:

Since the dtype of this fake_quant is `quint8`, the output range should be
from 0 to 255.  Fixing.  This should address the numerical inaccuracies with
sigmoid and hardsigmoid with `FixedQParamsFakeQuantize` attached compared
to their quantized counterparts.

In a future PR, might be safer to also make the activation functions
using `FixedQParamsFakeQuantize` to explicitly specify their expected
output range and zero_point.  Leaving that for later, as this bugfix
should be landed urgently.

Test Plan:

Manual script which gives low SQNR before this PR and high SQNR after
this PR: https://gist.github.com/vkuzo/9906bae29223da72b10d6b6aafadba42

https://github.com/pytorch/pytorch/pull/47376, which can be landed after
this, adds a proper test.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D24751497](https://our.internmc.facebook.com/intern/diff/D24751497)

[ghstack-poisoned]

vkuzo commented on torch/quantization/fake_quantize.py:
```diff
@@ -204,7 +204,7 @@ class FixedQParamsFakeQuantize(FakeQuantizeBase):
                  scale,
                  zero_point,
-                 dtype,
+                 dtype=torch.quint8,
```
note: adding this so the default for dtype aligns with the defaults for quant_min/quant_max
